fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/OnlineIndexerBuildGroupedCountIndexTest.java

```diff
@@ -0,0 +1,504 @@
+                        values.put(numValue2, count);
+                    }).join();
+            context.commit();
+            return values;
+        }
+    }
+
+    private static <M extends Message> Map<Tuple, M> byPrimaryKey(@Nonnull Collection<M> records) {
+        Map<Tuple, M> recordMap = Maps.newHashMapWithExpectedSize(records.size());
+        records.forEach(rec -> recordMap.put(PRIMARY_KEY.evaluateMessageSingleton(null, rec).toTuple(), rec));
+        return recordMap;
+    }
+
+    private static Map<Integer, Long> expectedCountByGroup(@Nonnull Collection<TestRecords1Proto.MySimpleRecord> records) {
+        Map<Integer, Long> values = new HashMap<>();
+        records.forEach(rec -> values.compute(rec.getNumValue2(), (numValue2, count) -> count == null ? 1L : (count + 1L)));
+        return values;
+    }
+
+    private void validateCountByGroup(@Nonnull Collection<TestRecords1Proto.MySimpleRecord> records) {
+        Map<Integer, Long> expected = expectedCountByGroup(records);
+        Map<Integer, Long> scanned = countByGroup();
+        assertEquals(expected, scanned);
+    }
+
+    private void rebuildGroupedCount(@Nonnull Collection<TestRecords1Proto.MySimpleRecord> recordsBefore,
+                                     @Nullable Collection<TestRecords1Proto.MyOtherRecord> otherRecordsBefore,
+                                     @Nullable String sourceIndex,
+                                     @Nonnull RecordsUpdater updater) {
+        openSimpleMetaData(baseGroupedHook());
+        try (FDBRecordContext context = openContext()) {
+            recordsBefore.forEach(recordStore::saveRecord);
+            if (otherRecordsBefore != null) {
+                otherRecordsBefore.forEach(recordStore::saveRecord);
+            }
+            context.commit();
+        }
+
+        openSimpleMetaData(withCountIndexHook());
+        OnlineIndexer.IndexingPolicy.Builder policyBuilder = OnlineIndexer.IndexingPolicy.newBuilder()
+                .setForbidRecordScan(sourceIndex != null)
+                .setIfDisabled(OnlineIndexer.IndexingPolicy.DesiredAction.REBUILD)
+                .setIfWriteOnly(OnlineIndexer.IndexingPolicy.DesiredAction.ERROR)
+                .setIfReadable(OnlineIndexer.IndexingPolicy.DesiredAction.ERROR);
+        if (sourceIndex != null) {
+            policyBuilder.setSourceIndex(sourceIndex);
+        }
+
+        Collection<TestRecords1Proto.MySimpleRecord> recordsAfter;
+
+        try (OnlineIndexer indexer = newIndexerBuilder()
+                .setIndex(COUNT_BY_NUM_VALUE_2)
+                .setIndexingPolicy(policyBuilder.build())
+                .setLimit(10)
+                .setMaxRetries(Integer.MAX_VALUE)
+                .setRecordsPerSecond(OnlineIndexer.DEFAULT_RECORDS_PER_SECOND * 100)
+                .build()) {
+            CompletableFuture<?> buildFuture = indexer.buildIndexAsync(true);
+            recordsAfter = updater.update(recordsBefore, otherRecordsBefore);
```

Comment:
There's definitely a risk that these won't really end up testing concurrent problems, at least with the numbers in the test, but I don't have a good solution, other than make a very large test that has a ton of records, takes hours to build, and randomly does all sorts of mutations while doing so, and then asserts that the result is correct. But, the record layer doesn't really have the infrastructure to do that kind of testing.